### ISSUE TITLE
Can the transpose-reshape operation in unwarping layer be replaced by pixel shuffle? 

### ISSUE THREAD

**Issue by proteus1991** [erezposner/Fast_Dense_Feature_Extraction#3]
Hi,

First, this is a great work that largely improves the processing speed for patch-based CNN. After reading the paper, I found the transpose-reshape operation might be replaced by the pixel shuffle that is provided by the PyTorch framework. Is it correct?

I'm looking forward to your reply.

Best,

Xiaohong

**Comment by erezposner** [comment]
Well, if you look closely at the source code for PixelShuffle you would find out there isn't much difference between my implementation and pixelshuffle. 

It is merely a matrix interlacing procedure. 

From  https://s0pytorch0org.icopy.site/docs/0.4.0/_modules/torch/nn/functional.html
```
def pixel_shuffle(input, upscale_factor):
    r"""Rearranges elements in a tensor of shape :math:`[*, C*r^2, H, W]` to a
    tensor of shape :math:`[C, H*r, W*r]`.

    See :class:`~torch.nn.PixelShuffle` for details.

    Args:
        input (Tensor): Input
        upscale_factor (int): factor to increase spatial resolution by

    Examples::

        >>> ps = nn.PixelShuffle(3)
        >>> input = torch.empty(1, 9, 4, 4)
        >>> output = ps(input)
        >>> print(output.size())
        torch.Size([1, 1, 12, 12])
    """
    batch_size, channels, in_height, in_width = input.size()
    channels //= upscale_factor ** 2

    out_height = in_height * upscale_factor
    out_width = in_width * upscale_factor

    input_view = input.contiguous().view(
        batch_size, channels, upscale_factor, upscale_factor,
        in_height, in_width)

    shuffle_out = input_view.permute(0, 1, 4, 2, 5, 3).contiguous()
    return shuffle_out.view(batch_size, channels, out_height, out_width)
```

The general idea for nd to nd transformation  usually involves using two operations:
- permute
- reshape

There is a great explanation here:
https://stackoverflow.com/questions/47977238/intuition-and-idea-behind-reshaping-4d-array-to-2d-array-in-numpy?noredirect=1&lq=1
